### PR TITLE
Add Ruby 3 support, update (most) dependencies, migrate to GH Actions

### LINKED ISSUE #115: Upgrade to fakefs 0.5.0
It includes [my fix for Ruby 2.1.0](https://github.com/defunkt/fakefs/commit/538c9060c3f4defc19db469174ae7ddf7cd98008), but seems to break some specs when I tried a quick upgrade.

Help is appreciated!

### LINKED ISSUE #165: maid daemon / watch rule throws exception
Watch rule in daemon mode throws an error. The syntax looks right as this is the sample rule. 
Repeat rule works fine. Any ideas what's wrong? Tested with Ruby 2.0.0, 2.2.0 and 2.4.1_1. Thanks

```
> maid daemon -r rulesD.rb
Logging actions to "/Users/me/.maid/maid.log"
E, [2017-06-30T10:56:33.636150 #36743] ERROR -- : run() in thread failed: wrong number of arguments (2 for 1):
 /Library/Ruby/Gems/2.0.0/gems/listen-2.10.1/lib/listen/adapter/base.rb:46:in `block (2 levels) in configure'
/Library/Ruby/Gems/2.0.0/gems/rb-fsevent-0.10.1/lib/rb-fsevent/fsevent.rb:75:in `call'
/Library/Ruby/Gems/2.0.0/gems/rb-fsevent-0.10.1/lib/rb-fsevent/fsevent.rb:75:in `run'
/Library/Ruby/Gems/2.0.0/gems/listen-2.10.1/lib/listen/adapter/darwin.rb:39:in `_run'
/Library/Ruby/Gems/2.0.0/gems/listen-2.10.1/lib/listen/adapter/base.rb:59:in `block in start'
/Library/Ruby/Gems/2.0.0/gems/listen-2.10.1/lib/listen/internals/thread_pool.rb:7:in `call'
/Library/Ruby/Gems/2.0.0/gems/listen-2.10.1/lib/listen/internals/thread_pool.rb:7:in `block in add'

called from:
 /Library/Ruby/Gems/2.0.0/gems/celluloid-0.16.0/lib/celluloid/calls.rb:26:in `public_send'
/Library/Ruby/Gems/2.0.0/gems/celluloid-0.16.0/lib/celluloid/calls.rb:26:in `dispatch'
/Library/Ruby/Gems/2.0.0/gems/celluloid-0.16.0/lib/celluloid/calls.rb:63:in `dispatch'
/Library/Ruby/Gems/2.0.0/gems/celluloid-0.16.0/lib/celluloid/cell.rb:60:in `block in invoke'
/Library/Ruby/Gems/2.0.0/gems/celluloid-0.16.0/lib/celluloid/cell.rb:71:in `block in task'
/Library/Ruby/Gems/2.0.0/gems/celluloid-0.16.0/lib/celluloid/actor.rb:357:in `block in task'
/Library/Ruby/Gems/2.0.0/gems/celluloid-0.16.0/lib/celluloid/tasks.rb:57:in `block in initialize'
/Library/Ruby/Gems/2.0.0/gems/celluloid-0.16.0/lib/celluloid/tasks/task_fiber.rb:15:in `block in create'
```

```
> ruby -v
ruby 2.0.0p648 (2015-12-16 revision 53162) [universal.x86_64-darwin16]
```

```
> maid version
0.7.0
```

rulesD.rb:
```
Maid.rules do
  watch '~/Pictures' do
    rule 'organize images by dimensions' do
      where_content_type(dir('~/Pictures/*'), 'image').each do |image|
        width, height = dimensions_px(image)
        move(image, mkdir("~/Pictures/#{width}x#{height}"))
      end
    end
  end
end
```

### LINKED ISSUE #175: ArgumentError: wrong number of arguments (given 2, expected 1)
This spec is failing on Ruby 2.3.8 and above.  Help is appreciated!

```
  1) Maid::Tools#find returns an array of all the files' names when no block is given
     Failure/Error: expect(@maid.find(@dir)).to match_array([@dir_expand_path, @file_expand_path])
     ArgumentError:
       wrong number of arguments (given 2, expected 1)
     # ./vendor/bundle/ruby/2.3.0/gems/fakefs-0.4.3/lib/fakefs/dir.rb:83:in `entries'
     # ./lib/maid/tools.rb:335:in `each'
     # ./lib/maid/tools.rb:335:in `to_a'
     # ./lib/maid/tools.rb:335:in `find'
     # ./spec/lib/maid/tools_spec.rb:371:in `block (3 levels) in <module:Maid>'
```

### LINKED ISSUE #176: Update to v0.8.0.beta.1
Thanks to @phylor, there are a bunch of new changes.  I'd like to release this as v0.8.0 after we can verify that things are still working.  The first part of that is releasing a beta, which just involves bumping some versions, etc.  It's been a while since I've done a release, so after those changes are made, I also need to make sure I can release to rubygems.org.

Help with bumping the internal versions and testing would be very appreciated!  I'd also be up for updating dependencies.  We can drop any old Ruby versions that stop working, but if the tests are still passing, I'm fine with keeping them for now.

I'm still using maid, but much less than I used to, so thank you @phylor for making some contributions.

### LINKED ISSUE #181: New install on Ruby 2.7.0 through rbenv.
I am new to ruby and project. When I finished the installation and ran I get following errors.
I ran
`rbnev install 2.7.0 && 
rbenv global 2.7.0
gem install maid`
and added ~/.maid/rules.rb 

> maid
~/.rbenv/versions/2.7.0/lib/ruby/gems/2.7.0/gems/mime-types-2.99.3/lib/mime/types/logger.rb:26: warning: `_1' is reserved for numbered parameter; consider another name
~/.rbenv/versions/2.7.0/lib/ruby/gems/2.7.0/gems/mime-types-2.99.3/lib/mime/types/logger.rb:26: warning: `_2' is reserved for numbered parameter; consider another name
~/.rbenv/versions/2.7.0/lib/ruby/gems/2.7.0/gems/mime-types-2.99.3/lib/mime/types/logger.rb:26: warning: `_3' is reserved for numbered parameter; consider another name

---------------------------
> maid clean --dry-run

Traceback (most recent call last):
	11: from ~/.rbenv/versions/2.7.0/bin/maid:23:in `<main>'
	10: from ~/.rbenv/versions/2.7.0/bin/maid:23:in `load'
	 9: from ~/.rbenv/versions/2.7.0/lib/ruby/gems/2.7.0/gems/maid-0.7.0/bin/maid:5:in `<top (required)>'
	 8: from ~/.rbenv/versions/2.7.0/lib/ruby/gems/2.7.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
	 7: from ~/.rbenv/versions/2.7.0/lib/ruby/gems/2.7.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	 6: from ~/.rbenv/versions/2.7.0/lib/ruby/gems/2.7.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	 5: from ~/.rbenv/versions/2.7.0/lib/ruby/gems/2.7.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	 4: from ~/.rbenv/versions/2.7.0/lib/ruby/gems/2.7.0/gems/maid-0.7.0/lib/maid/app.rb:50:in `clean'
	 3: from ~/.rbenv/versions/2.7.0/lib/ruby/gems/2.7.0/gems/maid-0.7.0/lib/maid/maid.rb:78:in `load_rules'
	 2: from ~/.rbenv/versions/2.7.0/lib/ruby/gems/2.7.0/gems/maid-0.7.0/lib/maid.rb:28:in `with_instance'
	 1: from ~/.rbenv/versions/2.7.0/lib/ruby/gems/2.7.0/gems/maid-0.7.0/lib/maid/maid.rb:82:in `block in load_rules'
~/.rbenv/versions/2.7.0/lib/ruby/gems/2.7.0/gems/maid-0.7.0/lib/maid/maid.rb:82:in `load': ~/.maid/rules.rb:21: syntax error, unexpected end-of-input, expecting `end' (SyntaxError)

### PR DESCRIPTION
**Pull request by coaxial**
In this PR:
- Ruby 3 support (3.0, 3.1, 3.2)
- Update (most) dependencies to latest available/compatible version
- Fix tests for Ruby 3
- Slightly refactor tests that messed the global state and caused failures when run all together (but not individually)
- Migrate to GH Actions

I've also updated the `Changelog` file but I'm not sure whether @benjaminoakes should be in it rather than myself.

I've bumped the gem version to 0.9.0 and updated all the metadata I could think of. I've also added the newer rubies to ~~the `.travis.yml` file~~ run on GH Actions.

I've made each gem update separate commits so that it's possible to `git bisect` should something break.

I have tested locally with Rubies 2.6, 2.7, 3.0, 3.1, 3.2 and running `rspec` to see 0 failures. 

Hope this gets merged as Ruby 2.6 and 2.7 is getting complicated to run on recent Ubuntu and derivatives because of an openssl issue (https://github.com/rvm/rvm/issues/5209).

@benjaminoakes if you're looking for a new maintainer, I'd be up for it. I use maid daily (and have for years) to keep my downloads, screenshots, and trash orderly; I find it super useful. 

@benjaminoakes I think you might need to enable GH Actions (https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#controlling-changes-from-forks-to-workflows-in-public-repositories) before the workflow can run on this PR.

Close #181
Close #176 
Close #165 (dependencies updated, no other reports)
Close #115

### CODE REVIEW

**Comment by benjaminoakes** [comment]
Oh, thank you!  Great to see this contribution!  I might have some time today to review.

**Comment by coaxial** [comment]
@benjaminoakes I think you might need to enable GH Actions (https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#controlling-changes-from-forks-to-workflows-in-public-repositories) before the workflow can run on this PR.

**Comment by benjaminoakes** [comment]
> @benjaminoakes if you're looking for a new maintainer, I'd be up for it. I use maid daily (and have for years) to keep my downloads, screenshots, and trash orderly; I find it super useful.

Thank you for the offer! I'll be in touch.

**Comment by benjaminoakes** [comment]
@coaxial I own [the "maid" organization](https://github.com/maid), which seems perfect for this situation.  I'm inviting you as an owner.  My current thinking is that it would make sense to move things over to that organization.  **Please feel empowered to move Maid forward in a positive way!**  How to give access to publishing on RubyGems is an open question still.  I'd like to give you all the access you need to make progress independently of me.

I'd still like to be _somewhat_ involved, but Maid has certainly **not** been a priority for me for quite some time.  (I first wrote it _12 years ago_, back before leading teams at work, getting married, owning a house, and having a kid, among other things!)  I have fond memories of working on Maid, so please be a good steward of the project.

**Review by benjaminoakes — APPROVED** [review]

**Comment by coaxial** [comment]
@benjaminoakes Thanks, I'll take good care of the project :)

This should help transfer the repo to the origanization: https://docs.github.com/en/repositories/creating-and-managing-repositories/transferring-a-repository#repository-transfers-and-organizations

File: ChangeLog
Comment (coaxial) [inline ChangeLog:1]: Do you have any rule of thumb to choose between `stable` and `unstable`?

File: .github/workflows/test.yml
Comment (benjaminoakes) [inline .github/workflows/test.yml:28]: Ruby 2.7 is the only Ruby listed here that's still supported (and [in two weeks](https://endoflife.date/ruby), only Ruby 3.x will be officially supported).  That said, there's _some_ value in making sure Maid works for older Rubies so that people can easily get started with `maid`.  Is supporting old Ruby versions worth the effort?  I leave that decision to you.

Comment (coaxial) [reply .github/workflows/test.yml:28]: I was thinking there might be legacy users still, and since the project was targetting 2.0 I wanted to try and keep that up.

But it largely depends on the GH Actions checks; if it fails for rubies < 2.7 then dropping support for them makes sense anyway since they're EOL (and probably complicated/impossible to install on currently supported OSes)

File: .github/workflows/test.yml
Comment (benjaminoakes) [inline .github/workflows/test.yml:9]: I don't seem to be able to get these builds to trigger.  That said, I ran the specs on a GitHub codespace and I can see that the specs pass for at least Ruby 3.1.3.  I'm opting to merge now so that this can move forward.  I hope to make a PR to test whether CI works, but I may leave that to you over on the Maid organization.

```
Finished in 1.19 seconds (files took 0.94519 seconds to load)
179 examples, 0 failures

$ ruby --version
ruby 3.1.3p185 (2022-11-24 revision 1a6b16756e) [x86_64-linux]
```

File: ChangeLog
Comment (benjaminoakes) [reply ChangeLog:1]: My process was always to do at least one unstable release before a stable release so that people could give feedback.  There is a list of users who have shared rules in [the wiki](https://github.com/benjaminoakes/maid/wiki).  Sometimes those users would help QA the releases and sometimes found issues, as I recall.

File: .github/workflows/test.yml
Comment (coaxial) [reply .github/workflows/test.yml:9]: There seems to be a syntax error in the file, which might be why it's not picking the workflow up. I'll look into it now.

**Comment by benjaminoakes** [comment]
As I mentioned earlier, I'm merging without passing specs.  Let's try to get those to trigger in a new PR.  I will also try following [the instructions you shared](https://docs.github.com/en/repositories/creating-and-managing-repositories/transferring-a-repository#repository-transfers-and-organizations) to move this project to a public organization.

File: ChangeLog
Comment (benjaminoakes) [reply ChangeLog:1]: You could look at what I did in old commits (it's been a while!), but I recall that once the `VERSION` became a stable version (i.e., change `0.9.0.alpha.1` to `0.9.0`), I would change the `ChangeLog` to read `stable` for that version.

File: .github/workflows/test.yml
Comment (benjaminoakes) [reply .github/workflows/test.yml:9]: Do you want a bit more time before I merge?  I was hoping to merge before trying to move this to the organization, just in case it doesn't go as smoothly as I might like.

Comment (coaxial) [reply .github/workflows/test.yml:9]: I fixed an indentation issue, maybe it shows in the repo's settings now? Otherwise, merge whenever you're happy with the PR and I'll take a look once I have access via the `maid` org.

Comment (benjaminoakes) [reply .github/workflows/test.yml:9]: > I fixed an indentation issue, maybe it shows in the repo's settings now?

I'm not seeing a difference right now.  I'm guessing it's some odd issue because we don't have any actions on `master`, but that's just a guess.

**Pull request merged**
